Add initial version of genesis deployment tool

hobofan opened the pull request:
Adds the first version of the genesis deployment script(s).

This allows us to generate the `accounts` section for a parity chainspec, taking as input the bytecode of the compiled contracts and their constructor arguments (+ ABI). This greatly simplifies the bootstrapping of a new Auxiliary chain.

Also adds `eslint` for JS style conformance as per #404.

**TODO**:
- [x] Find a way to be able to invoke script directly instead of via `truffle exec` (truffle exec has a poluted STDOUT output)
- [x] A bit simpler interface for adding Truffle contracts
- [x] Add script to merge the output of one of these deployment scripts into a chainspec
- [x] Allow for specifying a address for a contract instance instead of incrementing from a fixed address.
- [x] More robust way of incrementing addresses (that works for more than 255 addresses)
- [x] Better documentation of the `ContractRegistry` functions

@pgev : I gave you access to my fork

schemar commented on tools/deployment_tools.js:
```diff
@@ -0,0 +1,180 @@
+
+  takeNextAddress() {
+    const addressHex = web3.utils.bytesToHex(this.nextAvailableAddress);
+    this.nextAvailableAddress[this.nextAvailableAddress.length - 1] += 1;
```
Wouldn't it be easier to have a `BN` or similar and actually increase that? If I understand this approach correctly it is bound to 10 addresses, which may or may not be sufficient in the future.

hobofan replied:
It's bound to 255 (as it increments the last byte, not string character), but yeah, probably possible to use `BN` instead. I'll look into that.

schemar replied:
Doesn't it increment the last hex character (4 bits => max 16)? And does it know to go from `9` to `a`? JavaScript magic? 🤔

hobofan replied:
`this.nextAvailableAddress` is a byte array, so no magic there 😉

hobofan replied:
But that code is gone now anyway 😄

schemar replied:
Ah, somehow I thought it was a hex string 😳

hobofan commented:
From my side this would be mergeable in the current state. Should be in a usable state for @pgev's work on the ValidatorSet. Next step would be applying this to the deployment sequence for auxiliary, now that #448 is merged.

hobofan commented:
WIP again following some feedback from @pgev

hobofan commented:
Back to review.

@pgev I added the state machine we talked about. Also added some tests for core parts of the functionality and a referencing system for contract addresses in constructor arguments (so we can also order those correctly).

schemar commented on tools/deployment_tools.js:
```diff
@@ -0,0 +1,499 @@
+    generateAddress() {
+        const addressBytes = Web3.utils.sha3(rlp.encode([this.fromAddress, this.nextNonce]))
+            .slice(12)
+            .substring(14);
```
Should be extracted into a function with a meaningful name or alternatively explain in a comment what/why it does.

schemar commented on tools/deployment_tools.js:
```diff
@@ -0,0 +1,499 @@
+    /**
+     * Add a contract to the registry.
+     *
+     * @param {Contract} contracts Contract to add to the registry.
```
```suggestion
     * @param {Contract} contract Contract to add to the registry.
```

schemar commented on tools/deployment_tools.js:
```diff
@@ -0,0 +1,499 @@
+     * parity chainspec format.
+     *
+     * @param {Object} options.addressGenerator Address generator to use.
+     *                 Defaults to a IncrementingAddressGenerator.
```
```suggestion
     *                 Defaults to an IncrementingAddressGenerator.
```

schemar commented on tools/deployment_tools.js:
```diff
@@ -0,0 +1,499 @@
+        this.contracts.forEach(contract => contract.setAddress(addressGenerator));
+        this.contracts.forEach(contract => contract.instantiate());
+
+        const output = {
```
Better name? `genesisAccounts`?

schemar commented on tools/deployment_tools.js:
```diff
@@ -0,0 +1,499 @@
+     * This allows for deployment on a live network.
+     *
+     * @param {string} fromAddress See {@ IncrementingNonceAddressGenerator#constructor}.
+     * @param {number} startingNonce See {@ IncrementingNonceAddressGenerator#constructor}.
```
`@link`s?

hobofan replied:
Yes! 👍

schemar commented on tools/deployment_tools.js:
```diff
@@ -0,0 +1,499 @@
+    /**
+     * Generate transaction objects that can be passed as an argument
+     * to `web3.eth.sendTransaction()`.
+     * This allows for deployment on a live network.
```
It generates the objects *from this `ContractRegistry`,* right?
Maybe also explain that transactions and the parity genesis account are alternatives.

schemar commented on tools/deployment_tools.js:
```diff
@@ -0,0 +1,499 @@
+     * This is determined by the amount of transitive dependencies (including self).
+     */
+    _orderContracts() {
+        this.contracts.sort((a, b) => a._getDependenciesCount() - b._getDependenciesCount());
```
I don't understand how the dependencies count guarantees the order? Can't a contract with `n` dependencies depend on a contract with `n+1` or `n-1` dependencies? 😕

hobofan replied:
It's not just the count of dependencies but transitive dependencies. There is some graph theory behind it that @pgev can probably better explain 😅

schemar replied:
Ahhhh, yes. Now it makes sense 👍 I forgot to take the transitivity into account 😅

schemar commented on tools/merge_chainspec_accounts.js:
```diff
@@ -0,0 +1,19 @@
+#! /usr/bin/env node
+// Called like this:
+// node merge_chainspec_accounts.js chainspec.json accounts.json
```
you should also explain where chainspec and accounts comes from.

hobofan replied:
👍 
I'll add some docs for that script.

schemar commented on tools/merge_chainspec_accounts.js:
```diff
@@ -0,0 +1,19 @@
+  chainspec.accounts[address] = obj;
+});
+
+console.log(JSON.stringify(chainspec, null, 4));
```
What is the purpose of this?

schemar replied:
I don't know enough about JS, but are you using a *logging* mechanism to give feedback to the user on `stdout`? Can I change the behavior of `console.log()` to log somewhere else? E.g. a file? In that case I would prefer explicitly writing to `stdout` and not a log.

schemar commented on tools/test/deployment_tools.test.js:
```diff
@@ -0,0 +1,164 @@
+
+describe('Contract', () => {
+    describe('.setAddress()', () => {
+        test('correctly sets a fixed address for contract', () => {
```
We usually use mocha `it` as per truffle. Is that an option here?

hobofan replied:
We can definitely switch it to mocha + chai 👍 . I just picked Jest as I was most used to it and there wasn't a general JS testing framework in place.

schemar commented on tools/test/deployment_tools.test.js:
```diff
@@ -0,0 +1,164 @@
+
+            const output = registry.toParityGenesisAccounts();
+            expect(Object.keys(output)).toContain('0x0000000000444440000000000000000000000100');
+            expect(Object.keys(output).length).toEqual(1);
```
We usually use Chai `assert`s as per truffle. Is that an option here?

hobofan commented on tools/deployment_tools.js:
```diff
@@ -0,0 +1,499 @@
+const Web3 = require('web3');
+const fs = require('fs');
+const path = require('path');
+const rlp = require('rlp');
```
This should be according to JS best practices (which doesn't really prescribe either). The uppercase version of `Web3` is a bit unnormal, but it appears that this something that Ethereum JS community has settled on (as `web3` is used for an instance).

hobofan commented on tools/deploy_token_genesis.js:
```diff
@@ -0,0 +1,12 @@
+
+const EIP20Token = Contract.loadTruffleContract('EIP20Token', ['MYT', 'MyToken', 18]);
+EIP20Token.address = '0x0000000000444440000000000000000000000100';
+// EIP20Token.instantiate(['MYT', 'MyToken', 18]);
```
👍 Will delete. That was my manual test case before I added the tests.

hobofan commented on package.json:
```diff
@@ -6,9 +6,13 @@
+    "jest": "^23.6.0",
     "keccak": "1.4.0",
-    "rlp": "2.1.0",
+    "rlp": "^2.1.0",
```
👍 Will do.

Is there a specific reasoning behind it? Now that lockfiles are a thing in npm-land, I don't see much of a benefit there.

schemar replied:
In theory, I agree. However, the lock-file in npm behaves differently than what I would expect. For example, it gets changed when I run `npm install` (blows my mind).

Anyways, any update (even a bug-release), should always be a very conscious choice. That's pretty much the only reason.

hobofan commented on tools/deployment_tools.js:
```diff
@@ -0,0 +1,499 @@
+        const nextAddressBN = Web3.utils.toBN(addressHex).add(Web3.utils.toBN('1'));
+        this.nextAvailableAddress = `0x${Web3.utils.padLeft(nextAddressBN, 40)}`;
+
+        return addressHex;
```
No, it starts at `0x0000000000000000000000000000000000010000`

- saves address to variable `addressHex` (which is initially at `0x0000000000000000000000000000000000010000`)
- increments the internal variable used to track the next address
- returns `addressHex` which still is at `0x0000000000000000000000000000000000010000`

hobofan commented on tools/deployment_tools.js:
```diff
@@ -0,0 +1,499 @@
+
+        const contractPath = path.join(mergedOptions.rootDir, `build/contracts/${contractName}.json`);
+        const contents = fs.readFileSync(contractPath);
+        const truffleJson = JSON.parse(contents);
```
One good JS idiomatic alternative would be to have `loadTruffleContract` as a free-standing function, but I generally prefer keeping alternative constructors inside the class itself.

hobofan commented on tools/deployment_tools.js:
```diff
@@ -0,0 +1,499 @@
+     * @returns {Contract|null} The referenced contract.
+     */
+    _getReferenceContract(constructorArg) {
+        if (!((typeof constructorArg === 'object') && (constructorArg !== null)) || constructorArg.__type !== TYPE_CONTRACT_REFERENCE) {
```
I've tried a few things, but none of them were really an improvement. I'm open to suggestions!

schemar commented on tools/README.md:
```diff
@@ -0,0 +1,58 @@
+
+### Features
+
+- Can output the `accounts` section for a [Parity chainspec][parity-chainspec] to allow
```
```
 - Can output the `accounts` section for a [Parity chainspec][parity-chainspec] to allow
```
or
```suggestion
- Can output the `accounts` section for a [Parity chainspec](parity-chainspec) to allow
```

(compare line [43](https://github.com/OpenSTFoundation/mosaic-contracts/pull/458/files#diff-07b723ae57aaa00bce72a3022cf09ecdR43))

hobofan replied:
Should be []

schemar commented on tools/deployment_tool/contract.js:
```diff
@@ -0,0 +1,323 @@
+     * @returns {Contract|null} The referenced contract.
+     */
+    _getReferenceContract(constructorArg) {
+        if (!((typeof constructorArg === 'object') && (constructorArg !== null)) || constructorArg.__type !== TYPE_CONTRACT_REFERENCE) {
```
Maybe something like the below. I am sure this can inspire you to do something readable 😄 
```js
if(!isReference(constructorArg)) {
    return null;
}


function isReference(constructorArg) {
    if (!isInstance(constructorArg) {
        return false;
    }

    return constructorArg.__type === TYPE_CONTRACT_REFERENCE
}

function isInstance(constructorArg) {
    return typeof constructorArg === 'object' && constructorArg !== null;
}

```

hobofan replied:
👍

schemar commented on tools/deployment_tool/contract.js:
```diff
@@ -0,0 +1,323 @@
+        this._linkBytecode();
+
+        let constructorData = this.linkedBytecode();
+        constructorData += this._encodeConstructorArguments();
```
Mind adding a little bit of documentation to the above two lines? That would really help me a lot in understanding it.

schemar commented on tools/deployment_tool/contract_registry.js:
```diff
@@ -0,0 +1,127 @@
+    /**
+     * Generate transaction objects that can be passed as an argument
+     * to `web3.eth.sendTransaction()`.
+     * This allows for deployment on a live network.
```
It would help to reference `toParityGenesisAccounts` and `toLiveTransactionObjects` to each other I think. What do you think?

hobofan replied:
Done!

hobofan commented on package.json:
```diff
@@ -6,8 +6,13 @@
     "assert": "1.4.1",
     "bn.js": "4.11.8",
+    "chai": "4.2.0",
+    "eslint": "5.9.0",
```
Nope! Updating!

hobofan commented on tools/deployment_tool/contract.js:
```diff
@@ -0,0 +1,323 @@
+     *
+     * @param {string} contractName Name of the contract to load.
+     * @param {Array<*>} constructorArgs Arguments for the contract constructor.
+     * @param {object} options
```
Maybe.

schemar replied:
Then it could as well be a `rootDir` argument 🤷‍♂️

hobofan commented:
:repeat: :)

schemar commented on tools/deployment_tool/contract.js:
```diff
@@ -0,0 +1,323 @@
+    /**
+     * Returns the contract's address as detemined by the provided AddressGenerator.
+     *
+     * @param {Object|string} addressGeneratorOrAddress The AddressGenerator used for
```
Someone else should probably comment on this @benjaminbollen @pgev @deepesh-kn

deepesh-kn replied:
I too prefer separate functions like `setAddress` and `generateAddress`

deepesh-kn commented on tools/deployment_tool/contract.js:
```diff
@@ -0,0 +1,342 @@
+     * @param {string} contractName Name of the contract to create. If this contract is
+     *                 linked into another contracted, the provided contractName has to
+     *                 match the linking placeholder.
+     * @param {string} contracBytecode Bytecode of the contract.
```
Typo: `contracBytecode` -> `contractBytecode`

deepesh-kn commented on tools/deployment_tool/contract.js:
```diff
@@ -0,0 +1,342 @@
+    }
+
+    /**
+     * Returns the contract's address as detemined by the provided AddressGenerator.
```
Typo: `detemined` -> `determined`

deepesh-kn commented on tools/deployment_tool/test/deployment_tools.test.js:
```diff
@@ -0,0 +1,171 @@
+            registry.addContracts([MerklePatriciaProof, MessageBus]);
+
+            const output = registry.toParityGenesisAccounts();
+            assert.equal(Object.keys(output)[0], MerklePatriciaProof.getAddress());
```
We generally use strictEqual for the asserts.
Assert reason/messages are missing in all the assert statements.

hobofan commented:
@deepesh-kn @schemar

I addressed the feedback in the latest commit. `setAddress` has been split up into `setAddress` and `setGeneratedAddress`.
